File: julia/05-functions/higher_order.jl
# Higher-Order Functions in Julia

# Anonymous function assigned to a variable
double = x -> x * 2
println("double(21) = ", double(21))

numbers = [1, 2, 3, 4, 5, 6]

# map applies a function to every element
squared = map(x -> x^2, numbers)
println("squared = ", squared)

# filter keeps elements where the predicate is true
evens = filter(x -> x % 2 == 0, numbers)
println("evens = ", evens)

# reduce combines elements with a binary function
total = reduce(+, numbers)
println("sum = ", total)

# Functions can be passed as arguments
function apply_twice(f, x)
    return f(f(x))
end

println("apply_twice(double, 5) = ", apply_twice(double, 5))

# Closures: a function that returns a function, capturing `factor`
function multiplier(factor)
    return x -> x * factor
end

triple = multiplier(3)
println("triple(10) = ", triple(10))

# The do-block passes a multi-line function as the first argument
result = map([1, 2, 3]) do x
    x^2 + 1
end
println("do-block result = ", result)

File: python/05-functions/higher_order.py
# Higher-order functions, lambdas, and closures

def apply_twice(func, value):
    """Apply a function to a value two times."""
    return func(func(value))

double = lambda x: x * 2
print(apply_twice(double, 5))   # double(double(5))

# map and filter both take a function as their first argument
numbers = [1, 2, 3, 4, 5, 6]
squares = list(map(lambda n: n ** 2, numbers))
evens = list(filter(lambda n: n % 2 == 0, numbers))
print(squares)
print(evens)

# A closure: make_multiplier returns a function that remembers `factor`
def make_multiplier(factor):
    def multiplier(n):
        return n * factor
    return multiplier

triple = make_multiplier(3)
print(triple(10))
